zebra/debug_nl.c

```diff
@@ -0,0 +1,1291 @@
+		return "OIF";
+	case RTA_GATEWAY:
+		return "GATEWAY";
+	case RTA_PRIORITY:
+		return "PRIORITY";
+	case RTA_PREF:
+		return "PREF";
+	case RTA_PREFSRC:
+		return "PREFSRC";
+	case RTA_MARK:
+		return "MARK";
+	case RTA_METRICS:
+		return "METRICS";
+	case RTA_MULTIPATH:
+		return "MULTIPATH";
+	case RTA_PROTOINFO:
+		return "PROTOINFO";
+	case RTA_FLOW:
+		return "FLOW";
+	case RTA_CACHEINFO:
+		return "CACHEINFO";
+	case RTA_TABLE:
+		return "TABLE";
+	case RTA_MFC_STATS:
+		return "MFC_STATS";
+	case RTA_NH_ID:
+		return "NH_ID";
+	default:
+		return "UNKNOWN";
+	}
+}
+
+const char *neigh_rta2str(int type)
+{
+	switch (type) {
+	case NDA_UNSPEC:
+		return "UNSPEC";
+	case NDA_DST:
+		return "DST";
+	case NDA_LLADDR:
+		return "LLADDR";
+	case NDA_CACHEINFO:
+		return "CACHEINFO";
+	case NDA_PROBES:
+		return "PROBES";
+	case NDA_VLAN:
+		return "VLAN";
+	case NDA_PORT:
+		return "PORT";
+	case NDA_VNI:
+		return "VNI";
+	case NDA_IFINDEX:
+		return "IFINDEX";
+	case NDA_MASTER:
+		return "MASTER";
+	case NDA_LINK_NETNSID:
+		return "LINK_NETNSID";
+	default:
+		return "UNKNOWN";
+	}
+}
+
+const char *ifa_rta2str(int type)
+{
+	switch (type) {
+	case IFA_UNSPEC:
+		return "UNSPEC";
+	case IFA_ADDRESS:
+		return "ADDRESS";
+	case IFA_LOCAL:
+		return "LOCAL";
+	case IFA_LABEL:
+		return "LABEL";
+	case IFA_BROADCAST:
+		return "BROADCAST";
+	case IFA_ANYCAST:
+		return "ANYCAST";
+	case IFA_CACHEINFO:
+		return "CACHEINFO";
+	case IFA_MULTICAST:
+		return "MULTICAST";
+	case IFA_FLAGS:
+		return "FLAGS";
+	default:
+		return "UNKNOWN";
+	}
+}
+
+const char *nhm_rta2str(int type)
+{
+	switch (type) {
+	case NHA_UNSPEC:
+		return "UNSPEC";
+	case NHA_ID:
+		return "ID";
+	case NHA_GROUP:
+		return "GROUP";
+	case NHA_GROUP_TYPE:
+		return "GROUP_TYPE";
+	case NHA_BLACKHOLE:
+		return "BLACKHOLE";
+	case NHA_OIF:
+		return "OIF";
+	case NHA_GATEWAY:
+		return "GATEWAY";
+	case NHA_ENCAP_TYPE:
+		return "ENCAP_TYPE";
+	case NHA_ENCAP:
+		return "ENCAP";
+	case NHA_GROUPS:
+		return "GROUPS";
+	case NHA_MASTER:
+		return "MASTER";
+	default:
+		return "UNKNOWN";
+	}
+}
+
+#define FLAG_WRITE(flag, str)                                                  \
+	do {                                                                   \
+		if ((flags & (flag)) == 0)                                     \
+			break;                                                 \
+		if (!first) {                                                  \
+			wlen = snprintf(buf, buflen, ",");                     \
+			buf += (size_t)wlen;                                   \
+			buflen += (size_t)wlen;                                \
+		}                                                              \
+		wlen = snprintf(buf, buflen, (str));                           \
+		buf += (size_t)wlen;                                           \
+		buflen -= (size_t)wlen;                                        \
+		first = 0;                                                     \
+	} while (0)
```

Comment:
I hesitate to suggest it because I'm sure this was done for performance reasons. However, I think `strlcat` is probably the better option here:
- don't need to track all these length variables
- don't need to track buffer pointer
- works without needing `first`
- don't need to reference variables not defined within the macro scope, pass buffer size as an argument instead

If you have this dumping enabled you're agreeing to a perf hit anyway, I think it's worth the slightly increased cycle overhead to buy less code, more safety & more clarity.

Reply:
You are right, I replaced it with a static inline function + your suggestions.